File: src/main/java/sk/hudak/prco/eshop/MojaLekarenProductParser.java
package sk.hudak.prco.eshop;

import lombok.extern.slf4j.Slf4j;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import sk.hudak.prco.api.EshopUuid;
import sk.hudak.prco.api.ProductAction;
import sk.hudak.prco.parser.UnitParser;
import sk.hudak.prco.parser.impl.JSoupProductParser;
import sk.hudak.prco.utils.ConvertUtils;
import sk.hudak.prco.utils.JsoupUtils;
import sk.hudak.prco.utils.UserAgentDataHolder;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Objects;
import java.util.Optional;

@Slf4j
@Component
public class MojaLekarenProductParser extends JSoupProductParser {

    @Autowired
    public MojaLekarenProductParser(UnitParser unitParser, UserAgentDataHolder userAgentDataHolder) {
        super(unitParser, userAgentDataHolder);
    }

    @Override
    public EshopUuid getEshopUuid() {
        return EshopUuid.MOJA_LEKAREN;
    }

    @Override
    protected int parseCountOfPages(Document documentList) {
        Element select = documentList.select("p.pagination__part.pagination__part--page").first();
        if (select == null) {
            //FIXME move to parent as constants for one page only
            return 1;
        }
        return Integer.valueOf(select.children().last().previousElementSibling().previousElementSibling().text());
    }

    @Override
    protected List<String> parsePageForProductUrls(Document documentList, int pageNumber) {
        List<String> urls = new ArrayList<>();
        for (Element element : documentList.select("li[class=product]")) {
            Element aHref = element.child(0).child(0);
            String href = aHref.attr("href");
            urls.add(getEshopUuid().getProductStartUrl() + href);
        }
        return urls;
    }

    @Override
    protected Optional<String> parseProductNameFromDetail(Document documentDetailProduct) {
        return Optional.of(documentDetailProduct.select("body > main > div.detail-top.list > div > h1").first())
                .filter(Objects::nonNull)
                .map(Element::text);
    }

    @Override
    protected Optional<String> parseProductPictureURL(Document documentDetailProduct) {
        Elements select = documentDetailProduct.select("div.product__img > a > picture > img");
        Element first = select.first();
        if (first == null) {
            return Optional.empty();
        }
        return Optional.of(getEshopUuid().getProductStartUrl() + first.attr("data-srcset"));
    }

    @Override
    protected boolean isProductUnavailable(Document documentDetailProduct) {
        return JsoupUtils.notExistElement(documentDetailProduct, "#frm-addMultipleToBasket-form > button > meta");
    }

    @Override
    protected Optional<BigDecimal> parseProductPriceForPackage(Document documentDetailProduct) {
        return Optional.of(documentDetailProduct.select("span[itemprop=price]").first())
                .filter(Objects::nonNull)
                .map(Element::text)
                .map(ConvertUtils::convertToBigDecimal);
    }

    @Override
    protected Optional<ProductAction> parseProductAction(Document documentDetailProduct) {
        //TODO impl
        return Optional.empty();
    }

    @Override
    protected Optional<Date> parseProductActionValidity(Document documentDetailProduct) {
        //TODO impl
        return Optional.empty();
    }
}

File: src/main/java/sk/hudak/prco/eshop/pharmacy/MojaLekarenProductParser.kt
package sk.hudak.prco.eshop.pharmacy

import org.apache.commons.lang3.StringUtils
import org.jsoup.nodes.Document
import org.jsoup.nodes.Element
import org.springframework.stereotype.Component
import sk.hudak.prco.api.EshopUuid
import sk.hudak.prco.api.EshopUuid.MOJA_LEKAREN
import sk.hudak.prco.builder.SearchUrlBuilder
import sk.hudak.prco.parser.eshop.JSoupProductParser
import sk.hudak.prco.parser.unit.UnitParser
import sk.hudak.prco.utils.ConvertUtils
import sk.hudak.prco.utils.JsoupUtils
import sk.hudak.prco.utils.UserAgentDataHolder
import java.math.BigDecimal
import java.util.*
import java.util.Optional.ofNullable

@Component
class MojaLekarenProductParser(unitParser: UnitParser,
                               userAgentDataHolder: UserAgentDataHolder,
                               searchUrlBuilder: SearchUrlBuilder)
    : JSoupProductParser(unitParser, userAgentDataHolder, searchUrlBuilder) {

    override val eshopUuid: EshopUuid = MOJA_LEKAREN

    // koli pomalym odozvam davam na 15 sekund
    override val timeout: Int = TIMEOUT_15_SECOND

    override fun parseCountOfPages(documentList: Document): Int {
        val select = documentList.select("p.pagination__part.pagination__part--page").first() ?: return SINGLE_PAGE_ONE
        return Integer.valueOf(select.children().last().previousElementSibling().previousElementSibling().text())
    }

    override fun parseUrlsOfProduct(documentList: Document, pageNumber: Int): List<String> {
        val urls = ArrayList<String>()
        for (element in documentList.select("li[class=product]")) {
            val aHref = element.child(0).child(0)
            val href = aHref.attr("href")
            urls.add(eshopUuid.productStartUrl + href)
        }
        return urls
    }

    override fun parseProductNameFromDetail(documentDetailProduct: Document): String? {
        var element: Element? = documentDetailProduct.select("div.detail-top.list > div > h1").first()
        if (element == null) {
            element = documentDetailProduct.select("div > article > h1").first()
        }
        if (element == null) {
            element = documentDetailProduct.select("div.detail-top.list > div > article > h1").first()
        }

        return ofNullable(element)
                .map { it.text() }
                .filter { StringUtils.isNotBlank(it) }
                .orElse(null)
    }

    override fun parseProductPictureURL(documentDetailProduct: Document): String? {
        val first = documentDetailProduct.select("div.product__img > a > picture > img").first()
                ?: return null
        return Optional.of(eshopUuid.productStartUrl + first.attr("data-srcset")) .orElse(null)
    }

    override fun isProductUnavailable(documentDetailProduct: Document): Boolean {
        return JsoupUtils.notExistElement(documentDetailProduct, "#frm-addMultipleToBasket-form > button > meta")
    }

    override fun parseProductPriceForPackage(documentDetailProduct: Document): Optional<BigDecimal> {
        var element = documentDetailProduct.select("span[itemprop=price]").first()
        if (element != null) {
            val attr = element.attr("content")
            if (attr.isNotBlank()) {
                return ofNullable(ConvertUtils.convertToBigDecimal(attr))
            }
        }
        return Optional.of(element)
                .filter { Objects.nonNull(it) }
                .map { it.text() }
                .filter { StringUtils.isNotBlank(it) }
                .map { ConvertUtils.convertToBigDecimal(it) }
    }

}
